8285040: PPC64 intrinsics for divideUnsigned and remainderUnsigned methods in java.lang.Integer and java.lang.Long

TheRealMDoerr opened the pull request:
Add match rules for UDivI, UModI, UDivL, UModL as on x86 (JDK-8282221). PPC64 doesn't have DivMod instructions which can deliver both results at once.
Note: The x86 tests can currently not be extended to this platform because https://bugs.openjdk.java.net/browse/JDK-8280120 is not yet implemented.

vnkozlov reviewed (approved):
Make sense.

TheRealMDoerr commented:
Thanks for the quick review!

merykitty commented:
Honestly I don't quite understand why `UseDivMod` would make a difference here, without intrinsics, `Integer::divideUnsigned` is implemented as `(int)(toUnsignedLong(dividend) / toUnsignedLong(divisor))` and `Integer::remainderUnsigned` is implemented similarly, why does `toUnsignedLong(dividend) / toUnsignedLong(divisor)` not combined with `toUnsignedLong(dividend) % toUnsignedLong(divisor)` in the first place?

Thanks.

TheRealMDoerr commented:
I'm not sure if I got your question correctly. The description above contains the history what I tried (see the individual commits).
1. Implemented 4 intrinsics: UDivI, UModI, UDivL, UModL.
2. After that, I noticed performance regression in the Integer micro benchmark. So, I had removed the 2 Integer nodes again.
3. I noticed that testDivideRemainderUnsigned was relatively slow. Reason is that the UDivL MachNode got emitted twice: Once directly and once for the modulo operation (see expand rule). I fixed that by implementing the UseDivMod optimization for the unsigned nodes.
4. After having it implemented, the performance of the integer nodes was also better, so I have added the 2 nodes back.

TheRealMDoerr commented:
I believe that some register moves get generated in the old code due to conversions between Integer and Long which C2 can't optimize out. Performance improvement is not huge for Integer types.
The `UseDivMod` optimization is needed for unsigned types, because the signed types have it already and we would get a performance regression due to duplicated divide nodes in the benchmarks which need both (`testDivideRemainderUnsigned`).

RealLucy reviewed (approved):
Changes look good. 
Nice performance improvement!

TheRealMDoerr commented:
Thanks for the reviews!
/integrate
